### PR TITLE
docs: Updating docs for e2e tests

### LINKED ISSUE #1303: Doc: a better doc about how to write and run e2e test and unit test for Backend
# Feature request

## Please describe your feature

Recently when my partners and I tried to finish the the issues of backend, we always found troubles in writting unit test cases and e2e test cases. I have to say the big reason is that we are not skilled enough with the current testing framework, it would be better to add a detailed documentation to guide the newcomer on how to start writting  test cases and local debugging.

## Describe the solution you'd like

add a `backend-test.md`

### LINKED ISSUE #1581: docker-compose building apisix image is failing (for e2e backend test)
# Bug report

## Describe the bug
Currently, the compose file located in `/api/test/docker` directory needs a `Dockerfile-apisix` for spinning up two apisix docker containers.
https://github.com/apache/apisix-dashboard/blob/48e66cded1f4f7b67d37a9f1337fca492d6f44f5/api/test/docker/docker-compose.yaml#L128-L132

But in the mentioned directory there is no such file.
<pre>
.
├── apisix_config2.yaml
├── apisix_config.yaml
├── apisix_logs
├── docker-compose.yaml
├── Dockerfile
├── manager-api-conf.yaml
└── upstream.conf
</pre>
It should be present in the mentioned path as per the [docs](https://github.com/apache/apisix-dashboard/blob/master/docs/en/latest/back-end-e2e.md#start-with-docker-compose), resulting in a failure in starting up the required containers.

## How to Reproduce
```bash
cd api/test/docker


docker-compose build
#or 
docker-compose up
```

## Screenshots

image

## System information

Linux turing 5.8.0-44-generic #50~20.04.1-Ubuntu SMP Wed Feb 10 21:07:30 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

Thank you.

### PR DESCRIPTION
**Pull request by bisakhmondal**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues
closes #1303
closes #1581 

Changes made:
- New doc introducing how to write e2e & unit tests.
- Merged the `back-end-e2e.md` doc with `back-end-tests.md` along with updating the contents.
- a script to automate setting up the environment for e2e tests.
- 
Thanks :)

### CODE REVIEW

**Comment by bisakhmondal** [comment]
cc @nic-chen, @juzhiyuan to review when you have time :)

Guys, please suggest me a markdown linter, it seems mine is not working😄

**Review by nic-chen — COMMENTED** [review]
could we put the content of `back-end-e2e.md` into `back-end-tests.md`?

**Comment by bisakhmondal** [comment]
Thanks, @nic-chen for fixing the lint :) It's really helpful. Definitely! we can merge these docs together.

File: docs/en/latest/back-end-e2e.md [hunk 1]
```diff
@@ -82,23 +82,25 @@ This document describes how to use E2E test locally.
...
+        - 172.16.238.11:2379
+        - 172.16.238.12:2379
+   ```
 
 2. Use `docker-compose` to run services such as `manager-api`, `apisix`, `etcd` and `upstream-node`, run the command.
 
    ```sh
    cd /(Your apisix-dashboard folder path)/api/test/docker
+   # Download the apisix dockerfile
+   curl -o Dockerfile-apisix https://raw.githubusercontent.com/apache/apisix-docker/master/alpine/Dockerfile
```
Comment (starsz) [inline docs/en/latest/back-end-e2e.md:103]: Just want to know why we need to download the apisix dockerfile.

File: docs/en/latest/back-end-tests.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+---
+title: Backend Tests
```
Comment (starsz) [inline docs/en/latest/back-end-tests.md:2]: Can we put the license first?

**Review by starsz — COMMENTED** [review]
Pretty good.Thank you for your contribution.

File: docs/en/latest/back-end-tests.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    Headers       map[string]string     // Request headers. Include authorization header for secure routes.
+    ExpectStatus  int                   // Expected HTTP status code from the response
+    ExpectCode    int                   // Code generated by the host. Generally 0 for http.StatusOK.
+    ExpectMessage string                // The response message provided in the response by the host.
+    ExpectBody    interface{}           // The expected message body as a response.
+    Sleep         time.Duration //ms    // Cooldown period before making next request.
+}
+```
+
+Now to run a test use `RunTestCase(tc HttpTestCase)` or `testCaseCheck(tc HttpTestCase, t *testing.T)` for the `e2enew` or `e2e` respectively.
```
Comment (Jaycean) [inline docs/en/latest/back-end-tests.md:71]: At present, we are doing the existing E2E test to migrate to ginkgo. IMHO, it's better to add a note here to prompt users to write E2E tests using ginkgo as much as possible. After the migration is completed, the E2E test method can be removed here.

File: docs/en/latest/back-end-e2e.md (hunk 1 above)
Comment (bisakhmondal) [reply docs/en/latest/back-end-e2e.md:103]: The docker-compose file needs to build the apisix image from a dockerfile which is currently present at apisix-docker repository. For more details please go through issue #1581 . Thanks

File: docs/en/latest/back-end-tests.md (hunk 3 above)
Comment (bisakhmondal) [reply docs/en/latest/back-end-tests.md:71]: Hii. I had added the note in line [48](https://github.com/apache/apisix-dashboard/pull/1625/files#diff-c6ae044f165581d7c8c5df0dac40b063a1461f412acf700ea02c65ed06faa176R47-R49). Would that be sufficient or should I add it here too explicitly?
Thanks

File: docs/en/latest/back-end-tests.md (hunk 2 above)
Comment (bisakhmondal) [reply docs/en/latest/back-end-tests.md:2]: Hii. Sorry, I didn't get you. Do you mean adding the license before the title?

File: docs/en/latest/back-end-tests.md (hunk 3 above)
Comment (Jaycean) [reply docs/en/latest/back-end-tests.md:71]: No problem. Can we just write the ginkgo test method?
If E2E is used, the old method still needs to be migrated.
cc @starsz @nic-chen @bisakhmondal  Do you have any good suggestions？

File: docs/en/latest/back-end-tests.md (hunk 2 above)
Comment (starsz) [reply docs/en/latest/back-end-tests.md:2]: > Hii. Sorry, I didn't get you. Do you mean adding the license before the title?

Yeah. That's what I mean.

File: docs/en/latest/back-end-tests.md (hunk 3 above)
Comment (starsz) [reply docs/en/latest/back-end-tests.md:234]: IMO, we can even not describe e2e test. Only introduce how to write ginkgo test.
For that we are migrating all of our e2e tests to e2enew.

Comment (nic-chen) [reply docs/en/latest/back-end-tests.md:234]: > IMO, we can even not describe e2e test. Only introduce how to write ginkgo test.
> For that we are migrating all of our e2e tests to e2enew.

+1 , we could only introduce how to write ginkgo test.

Comment (gxthrj) [reply docs/en/latest/back-end-tests.md:234]: +1 Add a note here， only accept ginkgo test in new PR

File: docs/en/latest/back-end-tests.md (hunk 2 above)
Comment (nic-chen) [reply docs/en/latest/back-end-tests.md:2]: I think we don't need to move the title, other docs do the same.

File: docs/en/latest/back-end-tests.md (hunk 3 above)
Comment (bisakhmondal) [reply docs/en/latest/back-end-tests.md:234]: Nice!! Making the necessary changes then :)

**Review by nic-chen — APPROVED** [review]

File: docs/en/latest/back-end-tests.md (hunk 2 above)
Comment (juzhiyuan) [reply docs/en/latest/back-end-tests.md:2]: yes

**Review by juzhiyuan — APPROVED** [review]
LGTM

**Review by juzhiyuan — APPROVED** [review]
Don't forget to update this file: 

https://github.com/apache/apisix-dashboard/blob/429253ec8825397da04fac5e16340ce466db3ef4/docs/en/latest/config.json#L17

**Comment by nic-chen** [comment]
ping @starsz @gxthrj

**Review by starsz — APPROVED** [review]

**Review by juzhiyuan — APPROVED** [review]

**Review by Jaycean — APPROVED** [review]
LGTM. Thks.

**Review by bzp2010 — APPROVED** [review]

**Review by liuxiran — APPROVED** [review]

**Comment by nic-chen** [comment]
merged. thanks a lot. @bisakhmondal

**Comment by bisakhmondal** [comment]
😁 Thank you, everyone, for such wonderful quick reviews. V2.5 💪

**Pull request merged**